### DISCUSSION TITLE
How to authenticate with google

### DISCUSSION THREAD

**Discussion by RLD-JL** [supabase-community/supabase-kt#41] [Q&A]
I've done a setup and when trying to authenticate  with google it redirects me to this type of url if I use the `signUpWith ` I get `https://*********.supabase.co/auth/v1/auth/v1/authorize?provider=google&redirect_to=http://localhost:3001` and if I remove the extra `auth/v1` then it works.

If I use the `loginWith` then it is fine.

But I also don't know how to get a response/result of the  authentication because there is no onFail or onSuccess or something for me as shown in this example.

```
 client.gotrue.loginWith(Discord) {
        //optional:
        onFail = {
            when (it) {
                is OAuthFail.Timeout -> {
                    println("Timeout")
                }
            }
        }
        timeout = 50.seconds
        htmlTitle = "Supabase"
        htmlText = "Logged in. You may continue in the app."
    }
}
```

Here is a pic when I try to access onFail

![image](https://user-images.githubusercontent.com/106017010/209799609-1f35fd7b-b204-4da4-b223-d1d4aaf5f1a0.png)

**Comment by jan-tennert** [comment]
You mustn't specify auth/v1 in the supabase client builder. Which version are you using? And which platform?

**Comment by RLD-JL** [comment]
I did not specify the auth/v1 in the supabase client, using 0.7.2 and on web

**Comment by jan-tennert** [comment]
The code you showed is for Desktop. There is also an example for Web
![image](https://user-images.githubusercontent.com/26686035/209854661-83fa656a-f326-49ab-83fc-9c69bd18fb2d.png)

Also you see if the OAuth was successful if GoTrue#sessionStatus is Authenticated. You can collect the sessionStatus Flow to receive changes or use it as a state in Compose for Web

**Reply by jan-tennert** [reply]
I just updated the wiki for further clarification. Signing up with OAuth works like Login in with OAuth (Fail -> no session, Success -> valid session)

**Reply by jan-tennert** [reply]
And for the problem with the double 'auth/v1': Found the problem, publishing a fix

**Comment by sampatsharma143** [comment]
Hey @jan-tennert  i am trying to signup using email and password but getting this error.

`io.github.jan.supabase.exceptions.HttpRequestException: HTTP request to http://localhost?redirect_to=supabase%3A%2F%2Flogin (POST) failed with message: Operation not permitted`

i am not able to understand why i am getting this http://localhost 

`       client.gotrue.signUpWith(Email) {
                        email = findViewById<EditText>(R.id.email).text.toString()
                        password = findViewById<EditText>(R.id.password).text.toString()
                    }`


please help.

**Reply by jan-tennert** [reply]
> Hey @jan-tennert i am trying to signup using email and password but getting this error.
> 
> `io.github.jan.supabase.exceptions.HttpRequestException: HTTP request to http://localhost?redirect_to=supabase%3A%2F%2Flogin (POST) failed with message: Operation not permitted`
> 
> i am not able to understand why i am getting this http://localhost
> 
> ` client.gotrue.signUpWith(Email) { email = findViewById<EditText>(R.id.email).text.toString() password = findViewById<EditText>(R.id.password).text.toString() }`
> 
> please help.

If your supabase builder looks like this, it should be alright.
```kotlin
val client = createSupabaseClient(
     supabaseUrl = "https://id.supabase.co",
     supabaseKey = "key"
) {

    install(GoTrue)
    
}
```
 Do you have the internet Intent on android?

**Reply by sampatsharma143** [reply]
my code is exactly like this and by Internet intent do you mean internet permission in androidManifest.xml 
if yes thn , yeah i have it.

**Reply by sampatsharma143** [reply]
i have added these two dependancies in my gradle file :

```
    implementation("io.github.jan-tennert.supabase:gotrue-kt:0.9.0-beta-1")

    implementation 'io.ktor:ktor-client-cio:2.2.4'
```

**Reply by jan-tennert** [reply]
And you don't have a custom url in  the GoTrue config right? Also are you sure thats the code producing the error?

**Reply by sampatsharma143** [reply]
i am not sure this is my first time implementing supabase in android. 
I am not setting any custom url in GoTrue config. i even tried to add a custom url to check that if it changes that "http://localhost" but it didn't. 
am i implementing anything wrong?

**Reply by sampatsharma143** [reply]
also i have created 2 project on supabase website to test this and also setup supabase locally using docker. it is same.  as soon as i click on the button to call the signUpWith or loginWith this is the error that i get. 
`
D/KtorSupabaseHttpClient$request: Request failed with Operation not permitted
W/System.err: io.github.jan.supabase.exceptions.HttpRequestException: HTTP request to http://localhost?redirect_to=supabase%3A%2F%2Flogin (POST) failed with message: Operation not permitted
W/System.err:     at io.github.jan.supabase.network.KtorSupabaseHttpClient.request(KtorSupabaseHttpClient.kt:46)
W/System.err:     at io.github.jan.supabase.network.KtorSupabaseHttpClient$request$1.invokeSuspend(Unknown Source:15)
W/System.err:     at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
W/System.err:     at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
W/System.err:     at kotlinx.coroutines.EventLoop.processUnconfinedEvent(EventLoop.common.kt:69)
W/System.err:     at kotlinx.coroutines.internal.DispatchedContinuation.resumeWith(DispatchedContinuation.kt:355)
W/System.err:     at io.ktor.util.pipeline.SuspendFunctionGun.resumeRootWith(SuspendFunctionGun.kt:138)
W/System.err:     at io.ktor.util.pipeline.SuspendFunctionGun.access$resumeRootWith(SuspendFunctionGun.kt:11)
W/System.err:     at io.ktor.util.pipeline.SuspendFunctionGun$continuation$1.resumeWith(SuspendFunctionGun.kt:55)
W/System.err:     at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
W/System.err:     at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
W/System.err:     at android.os.Handler.handleCallback(Handler.java:883)
W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:100)
W/System.err:     at android.os.Looper.loop(Looper.java:214)
W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:7356)
W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
W/System.err:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
`

**Reply by jan-tennert** [reply]
No you are right, you don't need a custom url if you provide the supabase url in the client builder. The code snippet you showed me produces this error, right? And your app can connect to the internet? localhost is sometimes a fallback url

**Reply by sampatsharma143** [reply]
okay just tested it with real device it worked . i think the issue was internet connection

**Reply by jan-tennert** [reply]
Okay great, don't know why this is even happening because this error is really confusing,

**Reply by sampatsharma143** [reply]
yeah it confused me for 8 hours. i don't know if that url was supposed to came from the config file or from the internet .

**Reply by sampatsharma143** [reply]
now another is now that it is working, but client.gotrue.sessionStatus still returns not Authenticated . do you know why ?

**Reply by sampatsharma143** [reply]
I will try to fix it on my own but thank you for creating such wonderful library.

**Reply by jan-tennert** [reply]
If you don't have autoconfirm enabled, the sign up method doesn't automatically login the user. The user has to verify their email, and then you can log in using `client.gotrue.loginWith(Email)`.

**Reply by sampatsharma143** [reply]
yes finally got it. sorry to ask you this but is there any way that we can open the OAuth url ( google ) inside the app or by using the CustomTabs because , when i tried the google auth it opens in my phone browser and even after getting back to the app that google account url was still opened in my browser. so i think if we can open the link in customTab we will be able to close that after successfully login .

**Reply by jan-tennert** [reply]
Yea thats probably possible, but I didn't look into that. You can also use Google in-app sign in, but you have to setup the whole GoogleSignInClient yourself and request the idtoken. Then you can login in supabase like that:
```kotlin
client.gotrue.loginWith(IDToken) {
    idToken = ""
    clientId = "" //oauth client id I think
    provider = Google
}
```

**Reply by sampatsharma143** [reply]
okay i am ready to write the whole thing , it will be amazing i just hate that it opens up in a browser just for account selection.

**Reply by jan-tennert** [reply]
> okay i am ready to write the whole thing , it will be amazing i just hate that it opens up in a browser just for account selection.

yep as I said you can use Google in-app sign in, then you don't need any browser but you have to setup that up yourself and then put in the id token and client id in supabase.

**Reply by sampatsharma143** [reply]
I am gonna build that login system tomorrow. thanks again i was waiting for appwrite to add relational database or something that will help me to complete my project. but now i will switch to supabase . thanks again 😊

**Reply by purbojati** [reply]
Hey, how is it?
Do you mind to share it to public?
Thank you

**Reply by jan-tennert** [reply]
@purbojati If you are using Jetpack Compose, you can use https://github.com/supabase-community/supabase-kt/tree/master/plugins/ComposeAuth

**Comment by sampatsharma143** [comment]
Please add this in  GoTrue doc to add Internet permission in AndroidManifest.xml file. it is crucial

**Comment by jan-tennert** [comment]
> yes finally got it. sorry to ask you this but is there any way that we can open the OAuth url ( google ) inside the app or by using the CustomTabs because , when i tried the google auth it opens in my phone browser and even after getting back to the app that google account url was still opened in my browser. so i think if we can open the link in customTab we will be able to close that after successfully login .

There is now a [demo](https://github.com/supabase-community/supabase-kt/blob/development/demos/android-login/README.md) available on the development branch showcasing Native Google Sign In & OAuth using a In-App WebView
